### PR TITLE
[SPARK-45927][PYTHON] Update path handling for Python data source

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR updates how to handle `path` values from the `load()` method.
It changes the DataSource class constructor and add `path` as a key-value pair in the options field.

Also, this PR blocks loading multiple paths.

### Why are the changes needed?

To make the behavior consistent with the existing data source APIs.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @HyukjinKwon @cloud-fan

File: python/pyspark/sql/datasource.py [hunk 1]
```diff
@@ -45,30 +45,19 @@ class DataSource(ABC):
     """
 
     @final
-    def __init__(
-        self,
-        paths: List[str],
-        userSpecifiedSchema: Optional[StructType],
```
Comment (cloud-fan) [inline python/pyspark/sql/datasource.py:51]: why do we also remove user specified schema?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 2]
```diff
@@ -246,7 +246,15 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
...
     // Unless the legacy path option behavior is enabled, the extraOptions here
     // should not include "path" or "paths" as keys.
-    val plan = builder(sparkSession, source, paths, userSpecifiedSchema, extraOptions)
+    // Add path to the options field. Note currently it only supports a single path.
+    val optionsWithPath = if (paths.isEmpty) {
+      extraOptions
+    } else if (paths.length == 1) {
+        extraOptions + ("path" -> paths.head)
+    } else {
+      throw QueryCompilationErrors.multiplePathsUnsupportedError(source, paths)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:255]: This breaks all v1 sources, right? I think we should either follow v1 and ignore multi paths for python data source, or only apply this check for python data sources

File: python/pyspark/sql/datasource.py (hunk 1 above)
Comment (allisonwang-db) [reply python/pyspark/sql/datasource.py:51]: This field is actually not used. Both the `reader` and `writer` functions take in the `schema` parameter, and we can pass in the actual schema there.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 2 above)
Comment (allisonwang-db) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:255]: Yes we only apply this check to Python data sources (it's in `loadUserDefinedDataSource). The behavior is indeed different from the v1 source, but I find it more user-friendly to raise an explicit error than silently ignoring multiple paths.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 3]
```diff
@@ -246,7 +246,15 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
     val builder = sparkSession.sharedState.dataSourceManager.lookupDataSource(source)
     // Unless the legacy path option behavior is enabled, the extraOptions here
     // should not include "path" or "paths" as keys.
-    val plan = builder(sparkSession, source, paths, userSpecifiedSchema, extraOptions)
+    // Add path to the options field. Note currently it only supports a single path.
+    val optionsWithPath = if (paths.isEmpty) {
+      extraOptions
+    } else if (paths.length == 1) {
+        extraOptions + ("path" -> paths.head)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:253]:
```suggestion
      extraOptions + ("path" -> paths.head)
```

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:255]: does it help to add a `paths` option using JSON to hold String[]?

Comment (allisonwang-db) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:255]: Yea, let's just follow the DSv2 approach (options['paths'] = json serialized string list) to make Python data source behave the same as DSv2. I will update this.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonDataSourceSuite.scala [hunk 4]
```diff
@@ -161,12 +161,12 @@ class PythonDataSourceSuite extends QueryTest with SharedSparkSession {
       s"""
          |from pyspark.sql.datasource import DataSource, DataSourceReader
          |class SimpleDataSourceReader(DataSourceReader):
-         |    def __init__(self, paths, options):
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonDataSourceSuite.scala:164]: regardless we should remove this `paths`. Not all Python Datasources require `pathes`

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0. Thank you, @allisonwang-db and all.

**Review by HyukjinKwon — COMMENTED** [review]
LGTM2

**Pull request closed without merging**